### ISSUE TITLE
lambdas with common prefix produce unexpected results

### ISSUE THREAD

**Issue by jshook** [virtualdataset/virtdata-java#40]
This snippet is from a downstream project, but it illustrates the issue:

bindings:
 alpha: Identity()
blocks:
 - statements:
   - "{alpha} -> {a}, fake cycle {b} -> {c}\n"
   bindings:
    a: compose HashRange(0,2147483647) -> int
    b: HashRange(0,10)
    c: compose HashRange(0,10) -> long; HashRange(0,2147483647) -> int

0 -> 504793915, fake cycle 9 -> 1606844830
1 -> 1270630824, fake cycle 4 -> 1950522959
2 -> 32796775, fake cycle 3 -> 1280820171
3 -> 1556714733, fake cycle 3 -> 1280820171
4 -> 1992178435, fake cycle 1 -> 1606844830
5 -> 290580200, fake cycle 0 -> 1606844830
6 -> 907578887, fake cycle 6 -> 1829715121
7 -> 1905460236, fake cycle 2 -> 1280820171
8 -> 915325821, fake cycle 6 -> 1606844830
9 -> 853542728, fake cycle 2 -> 1606844830

initial inputs 7 and 9 to *a* yield value 2.
The initial value yielded for 2 is 32796775, although when taken as input to the second stage of 'c', produces a different value of 1280820171, or 1606844830.

This is at least counter intuitive, but it also seems like a bug. Additional lambda construction tracing needs to be provided at the minimum.

The first level issue is that with the same lambda construction, that the fake cycle 2 appears to produce differing results. Whether this is because the lambda for 'c' is not a strict composition around 'a' might be useful for explaining why, it is still not apparent to the user whether this is the case or not, or that they could be divergent in structure at all, given the naming. In the example, the lambda constructions are not exactly equivalent unless the components are exactly equivalent down to the signature (including return types) due to the fact that these functions are required to handle overflow with their own recycling logic.
It isn't clear yet whether this issue is because of the type signatures not matching, but that is the likely culprit.

It would make sense to have configurable overflow handling for a given lambda, such that if a data type were not wide enough to handle the inputs 'with support', then an error could be thrown. For some users this interpretation is more important, for others, the automatic recycling of datatype width is more important.

The second level issue is that the implied operators should yield self-consistent and obvious results. Regardless of whether there is an explanation for *why* the behavior is variant, the expectation that a simple mathematical construct should just work as is should be honored. This is a qualitative issue around user experience.

In short, this issue exposes multiple issues, some of them at the design level, some of them bugs, and maybe some in the lack of transparency of how this system works. It will probably result in multiple improvements once time is available.

**Comment by jshook** [comment]
Using the latest version, this recipe works as expected:
```
statements:
 - "x={x} f({x})={f} g({x})={g} g(f({x})={f})={gf}\n"
bindings:
 x: Identity()
 f: HashRange(0,10) -> long
 g: HashRange(0,2147483647) -> int
 gf: HashRange(0,10) -> long; HashRange(0,2147483647) -> int
```
output:
```
x=0 f(0)=9 g(0)=504793915 g(f(0)=9)=853542728
x=1 f(1)=4 g(1)=1270630824 g(f(1)=4)=1992178435
x=2 f(2)=3 g(2)=32796775 g(f(2)=3)=1556714733
x=3 f(3)=3 g(3)=1556714733 g(f(3)=3)=1556714733
x=4 f(4)=1 g(4)=1992178435 g(f(4)=1)=1270630824
x=5 f(5)=0 g(5)=290580200 g(f(5)=0)=504793915
x=6 f(6)=6 g(6)=907578887 g(f(6)=6)=907578887
x=7 f(7)=2 g(7)=1905460236 g(f(7)=2)=32796775
x=8 f(8)=6 g(8)=915325821 g(f(8)=6)=907578887
x=9 f(9)=2 g(9)=853542728 g(f(9)=2)=32796775
```
To relate to the above example with some implied currying syntax:
- f(x) is HashRange(0,10)(x)
- g(x) is HashRange(0,2147483647)(x)
- g(f(x)) is HashRange(0,2147483647)(HashRange(0,10)(x))

Specifically:
- f(7)=2 and f(9)=2
- g(2)=32796775
- thus g(f(7))==g(f(9))==32796775, which is correct

I think this can be closed out.
